text/0258-env-context-baggage-carriers.md

```diff
@@ -0,0 +1,236 @@
+# Environment Variable Specification for Context and Baggage Propagation
+
+This is a proposal to add Environment Variables to the OpenTelemetry
+specification as a carrier for context and baggage propagation between
+processes.
+
+## Table of Contents
+
+* [Motivation](#motivation)
+* [Explanation](#explanation)
+* [Internal Details](#internal-details)
+  * [UNIX](#unix)
+  * [Windows](#windows)
+  * [Allowed Characters](#allowed-characters)
+* [Trade-offs and Mitigations](#trade-offs-and-mitigations
+  * [Case-sensitivity](#case-sensitivity)
+  * [Security](#security)
+* [Prior Art and Alternatives](#prior-art-and-alternatives)
+  * [Alternatives and why they were not chosen](#alternatives-and-why-they-were-not-chosen)
+* [Open Questions](#open-questions)
+* [Future Possibilities](#future-possibilities)
+
+## Motivation
+
+The motivation for defining the specification for context and baggage
+propagation by using environment variables as carriers stems from a long open
+issue on the OpenTelemetry Specification repository, [issue #740](https://github.com/open-telemetry/opentelemetry-specification/issues/740#issue-665588273).
+
+The original issue identifies several use cases of systems that do not
+communicate across bounds by leveraging HTTP communications like, ETL, batch,
+and CI/CD systems, that need to benefit from context and baggage propagation.
+Defining a specification for this will enable better observability of these
+systems by enabling distributed tracing. There has already been a significant
+amount of [Prior Art](#prior-art) built within the industry and within
+OpenTelemetry to accomplish the immediate needs, however, OpenTelemetry at this
+time does not clearly define the specification for this form of propagation.
+
+Notably, as we define semantic conventions within the [CI/CD Working Group][cicd-wg],
+we'll need the specification defined for the industry to be able to adopt
+native tracing wtihin CI/CD systems.
+
+[cicd-wg]: https://github.com/open-telemetry/community/blob/main/projects/ci-cd.md
+
+## Explanation
+
+To propagate context and baggage between parent, and child processes
+in systems where HTTP communication does not occur, a specification for
+Environmental Variables can be used.
+
+Consider the following diagram:
+
+![Environment Variable Context Propagation](./img/0258-env-context-parent-child-process.png)
+
+In this diagram, a parent process is forked to spawn a child process,
+inheriting the environment variables from the original parent. The environment
+variables defined here, `TRACEPARENT`, `TRACESTATE`, and `BAGGAGE` are used to
```

Comment:
> The environment variables defined here

What does this phrase mean? Where is "here"? These concrete three names are what W3C Trace Context spec defines, they can be one possible implementation, e.g. if the SDK is configured with W3CPropagator. But the actual "env vars as carriers" should not care about which names are used, it should worry about the encoding of those names (and values) in such a way that the resulting output is safe for env vars (ideally across many OSs, ie. the lowest common denominator).

Reply:
This entire paragraph is discussing the example diagram above. "Here" means in that diagram. 

Are you saying that we should standardize on what the environment variable names are for context propagation? OpenTelemetry already has standards for environment variable names in its conventions. I don't understand why we wouldn't do the same here. Am I misunderstanding your statement?

Reply:
> OpenTelemetry already has standards for environment variable names in its conventions.

(a) they are used to configure SDK, not to affect per-request behavior, 

(b) we are looking to deprecate them in favor of configs, 

(c) the names of the metadata keys that constitute context/baggage are determined by the individual propagators. E.g. Jaeger propagator would use different HTTP headers than W3C Propagator. The same applies to env vars - there is no need to standardize on them, the names are already defined by propagators. But we need to standardize on encoding rules for them.

Reply:
> But the actual "env vars as carriers" should not care about which names are used

Agreed. We can recommend a few environment variables names for consistency purposes (e.g. BAGGAGE), but I think that should be all.

Reply:
>(a) they are used to configure SDK, not to affect per-request behavior,

AFAIU, the environment variables will affect any process that is running in the context where they are defined, so if you send and OpenTelemetry span (for example) in that context those requests should be affected by the context in the environment variables that will be the parent context of everything in that shell context (I say shell by simplicity).

Reply:
>(b) we are looking to deprecate them in favor of configs,

There is previous art in a few places, so IMO we should support the current environment variables in use. But it is not bad idea to allow use `OTEL_*` variants too.

Reply:
>(c) the names of the metadata keys that constitute context/baggage are determined by the individual propagators. E.g. Jaeger propagator would use different HTTP headers than W3C Propagator. The same applies to env vars - there is no need to standardize on them, the names are already defined by propagators. But we need to standardize on encoding rules for them.

We are talking about passing the OpenTelemetry parent context between processes, so it should not affect the propagation of HTTP headers.

For example, Suppose I run the application in the command line and pass a TRACEPARENT environment variable with the parent context related to my build in progress. In that case, all the spans and distributing tracing generated spans by that application will have that TRACEPARENT as the parent span. All the rest should work as work now in distributing tracing.

Reply:
I think this has been addressed through a couple iterations.